README.md
# Yaru Colors for Flutter

[![pub](https://img.shields.io/pub/v/yaru_colors.svg)](https://pub.dev/packages/yaru_colors)
[![CI](https://github.com/ubuntu/yaru_colors.dart/actions/workflows/ci.yaml/badge.svg)](https://github.com/ubuntu/yaru_colors.dart/actions/workflows/ci.yaml)
[![codecov](https://codecov.io/gh/ubuntu/yaru_colors.dart/branch/main/graph/badge.svg?token=xuGIiNAANl)](https://codecov.io/gh/ubuntu/yaru_colors.dart)

A collection of Yaru color palettes for Flutter.

```dart
import 'package:yaru_colors/yaru_colors.dart';
```

```dart
MyWidget(color: YaruColors.purple)
```

![accent](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/images/accent.png "accent colors")
![flavor](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/images/flavor.png "flavor colors")
![neutral](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/images/neutral.png "neutral colors")
![custom](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/images/custom.png "custom color")

# Usage

To use this package, add `yaru_colors` as a
[dependency in your pubspec.yaml file](https://docs.flutter.dev/development/packages-and-plugins/using-packages).

# References

- [yaru_color_generator](https://pub.dev/packages/yaru_color_generator) - Generates full Yaru Material color palettes from singular primary color values.
- [yaru_color_builder](https://pub.dev/packages/yaru_color_builder) - A builder extension for generating Yaru colors at build time.


CHANGELOG.md
## 0.1.7

* CI: pin Melos at 2.9.0 by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/29
* Sync accent colors with ubuntu/yaru by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/28

## 0.1.6

*  Revert jet color change for now by @Feichtmeier in https://github.com/ubuntu/yaru_colors.dart/pull/24

## 0.1.5

* Darken jet by @Feichtmeier in https://github.com/ubuntu/yaru_colors.dart/pull/24

## 0.1.4

* CI: replace 18.04 with 22.04 by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/15
* Restructure: promote `yaru_colors` to the root by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/16
* Docs: fix screenshot links by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/17
* CI: merge workflows by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/18
* Example: get rid of the backwards yaru.dart dependency by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/20
* Add YaruColors.ubuntuCinnamonBrown by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/19
* Add Unity purple by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/21
* Import generated butterfly pink by @jpnurmi in https://github.com/ubuntu/yaru_colors.dart/pull/22

## 0.1.3

- Fix Flutter 3.1 deprecated member use, by @Jupi007 in ([#11](https://github.com/ubuntu/yaru_colors.dart/pull/11))
- melos: choose between dart test vs. flutter test, by @jpnurmi in ([#13](https://github.com/ubuntu/yaru_colors.dart/pull/13))
- Add luminosity Color extension, by @Jupi007 in ([#12](https://github.com/ubuntu/yaru_colors.dart/pull/12))

## 0.1.2

- Add YaruColors.titleBarLight/Dark ([#9](https://github.com/ubuntu/yaru_colors.dart/pull/9))

## 0.1.1

- Ubuntu Butterfly Pink ([#5](https://github.com/ubuntu/yaru_colors.dart/pull/5))

## 0.1.0

- Initial version.


test/yaru_color_extension_test.dart
import 'package:flutter/material.dart';
import 'package:test/test.dart';
import 'package:yaru_colors/src/yaru_color_extension.dart';

final Matcher throwsAssertionError = throwsA(isA<AssertionError>());
final Color midColor = const HSLColor.fromAHSL(.5, 180, .5, .5).toColor();

void main() {
  group('Color.scale() test -', () {
    test('With out of range amount', () {
      expect(() => midColor.scale(alpha: -1.1), throwsAssertionError);
      expect(() => midColor.scale(alpha: 1.1), throwsAssertionError);
      expect(() => midColor.scale(hue: -1.1), throwsAssertionError);
      expect(() => midColor.scale(hue: 1.1), throwsAssertionError);
      expect(() => midColor.scale(saturation: -1.1), throwsAssertionError);
      expect(() => midColor.scale(saturation: 1.1), throwsAssertionError);
      expect(() => midColor.scale(lightness: -1.1), throwsAssertionError);
      expect(() => midColor.scale(lightness: 1.1), throwsAssertionError);
    });
    test('With clamped amount', () {
      expect(
        midColor.scale(alpha: -1.0),
        const Color(0x0040bfbf),
      );
      expect(
        midColor.scale(alpha: 1.0),
        const Color(0xff40bfbf),
      );
      expect(
        midColor.scale(hue: -1.0),
        const Color(0x80bf4040),
      );
      expect(
        midColor.scale(hue: 1.0),
        const Color(0x80bf4040),
      );
      expect(
        midColor.scale(saturation: -1.0),
        const Color(0x80808080),
      );
      expect(
        midColor.scale(saturation: 1.0),
        const Color(0x8000ffff),
      );
      expect(
        midColor.scale(lightness: -1.0),
        const Color(0x80000000),
      );
      expect(
        midColor.scale(lightness: 1.0),
        const Color(0x80ffffff),
      );
    });
    test('With medium amount', () {
      expect(
        midColor.scale(alpha: -0.5),
        const Color(0x4040bfbf),
      );
      expect(
        midColor.scale(alpha: 0.5),
        const Color(0xc040bfbf),
      );
      expect(
        midColor.scale(hue: -0.5),
        const Color(0x8080bf40),
      );
      expect(
        midColor.scale(hue: 0.5),
        const Color(0x808040bf),
      );
      expect(
        midColor.scale(saturation: -0.5),
        const Color(0x80609f9f),
      );
      expect(
        midColor.scale(saturation: 0.5),
        const Color(0x8020dfdf),
      );
      expect(
        midColor.scale(lightness: -0.5),
        const Color(0x80206060),
      );
      expect(
        midColor.scale(lightness: 0.5),
        const Color(0x80a0dfdf),
      );
    });
  });

  group('Color.adjust() test -', () {
    test('With out of range amount', () {
      expect(() => midColor.adjust(alpha: -1.1), throwsAssertionError);
      expect(() => midColor.adjust(alpha: 1.1), throwsAssertionError);
      expect(() => midColor.adjust(hue: -360.1), throwsAssertionError);
      expect(() => midColor.adjust(hue: 360.1), throwsAssertionError);
      expect(() => midColor.adjust(saturation: -1.1), throwsAssertionError);
      expect(() => midColor.adjust(saturation: 1.1), throwsAssertionError);
      expect(() => midColor.adjust(lightness: -1.1), throwsAssertionError);
      expect(() => midColor.adjust(lightness: 1.1), throwsAssertionError);
    });
    test('With clamped amount', () {
      expect(
        midColor.adjust(alpha: -1.0),
        const Color(0x0040bfbf),
      );
      expect(
        midColor.adjust(alpha: 1.0),
        const Color(0xff40bfbf),
      );
      expect(
        midColor.adjust(hue: -180),
        const Color(0x80bf4040),
      );
      expect(
        midColor.adjust(hue: 180),
        const Color(0x80bf4040),
      );
      expect(
        midColor.adjust(saturation: -1.0),
        const Color(0x80808080),
      );
      expect(
        midColor.adjust(saturation: 1.0),
        const Color(0x8000ffff),
      );
      expect(
        midColor.adjust(lightness: -1.0),
        const Color(0x80000000),
      );
      expect(
        midColor.adjust(lightness: 1.0),
        const Color(0x80ffffff),
      );
    });
    test('With medium amount', () {
      expect(
        midColor.adjust(alpha: -0.25),
        const Color(0x4040bfbf),
      );
      expect(
        midColor.adjust(alpha: 0.25),
        const Color(0xc040bfbf),
      );
      expect(
        midColor.adjust(hue: -90),
        const Color(0x8080bf40),
      );
      expect(
        midColor.adjust(hue: 90),
        const Color(0x808040bf),
      );
      expect(
        midColor.adjust(saturation: -0.25),
        const Color(0x80609f9f),
      );
      expect(
        midColor.adjust(saturation: 0.25),
        const Color(0x8020dfdf),
      );
      expect(
        midColor.adjust(lightness: -0.25),
        const Color(0x80206060),
      );
      expect(
        midColor.adjust(lightness: 0.25),
        const Color(0x80a0dfdf),
      );
    });
  });

  group('Color.copyWith() test -', () {
    test('With out of range amount', () {
      expect(() => midColor.copyWith(alpha: -1.1), throwsAssertionError);
      expect(() => midColor.copyWith(alpha: 1.1), throwsAssertionError);
      expect(() => midColor.copyWith(hue: -360.1), throwsAssertionError);
      expect(() => midColor.copyWith(hue: 360.1), throwsAssertionError);
      expect(() => midColor.copyWith(saturation: -1.1), throwsAssertionError);
      expect(() => midColor.copyWith(saturation: 1.1), throwsAssertionError);
      expect(() => midColor.copyWith(lightness: -1.1), throwsAssertionError);
      expect(() => midColor.copyWith(lightness: 1.1), throwsAssertionError);
    });
    test('With various amount', () {
      expect(
        midColor.copyWith(alpha: 0.25),
        const Color(0x4040bfbf),
      );
      expect(
        midColor.copyWith(hue: 90),
        const Color(0x8080bf40),
      );
      expect(
        midColor.copyWith(saturation: 0.75),
        const Color(0x8020dfdf),
      );
      expect(
        midColor.copyWith(lightness: 0.75),
        const Color(0x80a0dfdf),
      );
    });
  });
}


example/integration_test/screenshot_test.dart
import 'dart:io';

import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';
import 'package:package_config/package_config.dart';
import 'package:path/path.dart' as p;
import 'package:yaru_colors_example/main.dart' as app;

// NOTE: run `flutter test --update-goldens integration_test` to update screenshots

Future<void> main() async {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets('accents', (tester) async {
    app.main();
    await tester.pumpAndSettle();

    const colors = [
      'orange',
      'olive',
      'bark',
      'sage',
      'prussianGreen',
      'viridian',
      'purple',
      'red',
      'blue',
      'magenta',
    ];

    for (final color in colors) {
      final swatch = find.byKey(Key('YaruColors.$color'), skipOffstage: false);
      await tester.ensureVisible(swatch);
      await tester.pump();
      await expectLater(swatch, matchesGoldenFile(await getImagePath(color)));
    }
  });

  testWidgets('flavors', (tester) async {
    app.main();
    await tester.pumpAndSettle();
    await tester.tap(find.text('Flavor'));
    await tester.pumpAndSettle();

    const colors = [
      'kubuntuBlue',
      'lubuntuBlue',
      'ubuntuBudgieBlue',
      'ubuntuButterflyPink',
      'ubuntuCinnamonBrown',
      'ubuntuMateGreen',
      'ubuntuStudioBlue',
      'ubuntuUnityPurple',
      'xubuntuBlue',
    ];

    for (final color in colors) {
      final swatch = find.byKey(Key('YaruColors.$color'), skipOffstage: false);
      await tester.ensureVisible(swatch);
      await tester.pump();
      await expectLater(swatch, matchesGoldenFile(await getImagePath(color)));
    }
  });

  testWidgets('neutral', (tester) async {
    app.main();
    await tester.pumpAndSettle();
    await tester.tap(find.text('Neutral'));
    await tester.pumpAndSettle();

    const colors = [
      'warmGrey',
    ];

    for (final color in colors) {
      final swatch = find.byKey(Key('YaruColors.$color'), skipOffstage: false);
      await tester.ensureVisible(swatch);
      await tester.pump();
      await expectLater(swatch, matchesGoldenFile(await getImagePath(color)));
    }
  });
}

// yaru_colors.dart/packages/yaru_colors/doc/images
Future<String> getImagePath(String name) async {
  final config = await findPackageConfig(Directory.current);
  final package = config?.packages.firstWhere((p) => p.name == 'yaru_colors');
  return p.join(package!.root.path, 'doc', 'images', '$name.png');
}


lib/src/colors.dart
import 'package:flutter/material.dart';
import 'package:yaru_color_generator/yaru_color_generator.dart';

part 'colors.g.dart';

/// Available Yaru color palettes.
class YaruColors {
  YaruColors._();

  /// Generates a Yaru color palette from the specicied [primary] color value.
  static MaterialColor generate(int primary) {
    final color = generateYaruColor(primary);
    return MaterialColor(color.primary, {
      for (final shade in color.shades) shade: Color(color[shade]!),
    });
  }

  /// Ubuntu Orange
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/orange.png)
  @GeneratedYaruColor(0xFFE95420)
  static const MaterialColor orange = _$YaruColors.orange;

  /// Warm Grey
  ///
  /// For balance. The addition of warm grey softens the combination of orange
  /// and aubergine and provides a bridge between the two.
  ///
  /// Warm grey can be used for; backgrounds, graphics, dot patterns, charts and
  /// diagrams. It can also be used for large size text.
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/warmGrey.png)
  @GeneratedYaruColor(0xFFAEA79F)
  static const MaterialColor warmGrey = _$YaruColors.warmGrey;

  /// Cool Grey
  ///
  /// For typography, particularly body copy. Black can be quite harsh in
  /// combination with aubergine, but grey delivers more balance while still
  /// being legible.
  ///
  /// Cool grey can also be used within charts and diagrams.
  static const Color coolGrey = Color(0xFF333333);

  /// Text Grey
  ///
  /// Text grey is used for small size headings, sub-headings and body copy text
  /// only.
  static const Color textGrey = Color(0xFF111111);

  /// Error
  static const Color error = Color(0xFFff0000);

  /// Warning
  static const Color warning = Color(0xFFf99b11);

  /// Success
  static const Color success = Color(0xFF0e8420);

  /// Porcelain
  static const Color porcelain = Color(0xFFFAFAFA);

  /// Inkstone
  static const Color inkstone = Color(0xFF3B3B3B);

  /// Jet
  static const Color jet = Color(0xFF2B2B2B);

  /// Light title bar
  static const Color titleBarLight = Color(0xFFEBEBEB);

  /// Dark title bar
  static const Color titleBarDark = Color(0xFF303030);

  /// Olive
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/olive.png)
  @GeneratedYaruColor(0xFF4B8501)
  static const MaterialColor olive = _$YaruColors.olive;

  /// Bark
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/bark.png)
  @GeneratedYaruColor(0xFF787859)
  static const MaterialColor bark = _$YaruColors.bark;

  /// Sage
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/sage.png)
  @GeneratedYaruColor(0xFF657B69)
  static const MaterialColor sage = _$YaruColors.sage;

  /// Prussian Green
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/prussianGreen.png)
  @GeneratedYaruColor(0xFF308280)
  static const MaterialColor prussianGreen = _$YaruColors.prussianGreen;

  /// Viridian
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/viridian.png)
  @GeneratedYaruColor(0xFF03875B)
  static const MaterialColor viridian = _$YaruColors.viridian;

  /// Purple
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/purple.png)
  @GeneratedYaruColor(0xFF7764D8)
  static const MaterialColor purple = _$YaruColors.purple;

  /// Red
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/red.png)
  @GeneratedYaruColor(0xFFDA3450)
  static const MaterialColor red = _$YaruColors.red;

  /// Blue
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/blue.png)
  @GeneratedYaruColor(0xFF0073E5)
  static const MaterialColor blue = _$YaruColors.blue;

  /// Magenta
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/magenta.png)
  @GeneratedYaruColor(0xFFB34CB3)
  static const MaterialColor magenta = _$YaruColors.magenta;

  /// Kubuntu Blue
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/kubuntuBlue.png)
  @GeneratedYaruColor(0xFF0079C1)
  static const MaterialColor kubuntuBlue = _$YaruColors.kubuntuBlue;

  /// Lubuntu Blue
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/lubuntuBlue.png)
  @GeneratedYaruColor(0xFF0068C8)
  static const MaterialColor lubuntuBlue = _$YaruColors.lubuntuBlue;

  /// Ubuntu Budgie Blue
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/ubuntuBudgieBlue.png)
  @GeneratedYaruColor(0xFF2196F3)
  static const MaterialColor ubuntuBudgieBlue = _$YaruColors.ubuntuBudgieBlue;

  /// Ubuntu Butterfly Pink
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/ubuntuButterflyPink.png)
  @GeneratedYaruColor(0xFFFF135B)
  static const MaterialColor ubuntuButterflyPink =
      _$YaruColors.ubuntuButterflyPink;

  /// Ubuntu Cinnamon Brown
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/ubuntuCinnamonBrown.png)
  @GeneratedYaruColor(0xFFDD682A)
  static const MaterialColor ubuntuCinnamonBrown =
      _$YaruColors.ubuntuCinnamonBrown;

  /// Ubuntu MATE Green
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/ubuntuMateGreen.png)
  @GeneratedYaruColor(0xFF73AF59)
  static const MaterialColor ubuntuMateGreen = _$YaruColors.ubuntuMateGreen;

  /// Ubuntu Studio Blue
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/ubuntuStudioBlue.png)
  @GeneratedYaruColor(0xFF009BF9)
  static const MaterialColor ubuntuStudioBlue = _$YaruColors.ubuntuStudioBlue;

  /// Ubuntu Unity Purple
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/ubuntuUnityPurple.png)
  @GeneratedYaruColor(0xFF9005D5)
  static const MaterialColor ubuntuUnityPurple = _$YaruColors.ubuntuUnityPurple;

  /// Xubuntu Blue
  ///
  /// ![](https://raw.githubusercontent.com/ubuntu/yaru_colors.dart/main/doc/images/xubuntuBlue.png)
  @GeneratedYaruColor(0xFF0044AA)
  static const MaterialColor xubuntuBlue = _$YaruColors.xubuntuBlue;
}


example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "Yaru Colors");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "Yaru Colors");
  }

  GdkGeometry geometry;
  geometry.min_width = 820;
  geometry.min_height = 640;
  gtk_window_set_geometry_hints(window, nullptr, &geometry, GDK_HINT_MIN_SIZE);
  gtk_window_set_default_size(window, 960, 680);

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(
      project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_show(GTK_WIDGET(window));
  gtk_widget_show(GTK_WIDGET(view));
  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application,
                                                  gchar*** arguments,
                                                  int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
    g_warning("Failed to register: %s", error->message);
    *exit_status = 1;
    return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line =
      my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID, "flags",
                                     G_APPLICATION_NON_UNIQUE, nullptr));
}


example/lib/main.dart
import 'package:flex_color_picker/flex_color_picker.dart';
import 'package:flutter/material.dart';
import 'package:yaru_colors/yaru_colors.dart';

const tileWidth = 360.0;
const tileHeight = 40.0;
const tileSpacing = 24.0;

const accentColors = {
  'YaruColors.orange': YaruColors.orange,
  'YaruColors.bark': YaruColors.bark,
  'YaruColors.sage': YaruColors.sage,
  'YaruColors.olive': YaruColors.olive,
  'YaruColors.viridian': YaruColors.viridian,
  'YaruColors.prussianGreen': YaruColors.prussianGreen,
  'YaruColors.blue': YaruColors.blue,
  'YaruColors.purple': YaruColors.purple,
  'YaruColors.magenta': YaruColors.magenta,
  'YaruColors.red': YaruColors.red,
};

const flavorColors = {
  'YaruColors.kubuntuBlue': YaruColors.kubuntuBlue,
  'YaruColors.lubuntuBlue': YaruColors.lubuntuBlue,
  'YaruColors.ubuntuBudgieBlue': YaruColors.ubuntuBudgieBlue,
  'YaruColors.ubuntuButterflyPink': YaruColors.ubuntuButterflyPink,
  'YaruColors.ubuntuCinnamonBrown': YaruColors.ubuntuCinnamonBrown,
  'YaruColors.ubuntuMateGreen': YaruColors.ubuntuMateGreen,
  'YaruColors.ubuntuStudioBlue': YaruColors.ubuntuStudioBlue,
  'YaruColors.ubuntuUnityPurple': YaruColors.ubuntuUnityPurple,
  'YaruColors.xubuntuBlue': YaruColors.xubuntuBlue,
};

const neutralColors = {
  'YaruColors.warmGrey': YaruColors.warmGrey,
};

const shades = [50, 100, 200, 300, 400, 500, 600, 700, 800, 900];

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  var _currentPage = 0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.light(useMaterial3: true),
      darkTheme: ThemeData.dark(useMaterial3: true),
      debugShowCheckedModeBanner: false,
      home: Builder(builder: (context) {
        return Scaffold(
          body: IndexedStack(
            index: _currentPage,
            children: const [
              ColorView(colors: accentColors),
              ColorView(colors: flavorColors),
              ColorView(colors: neutralColors),
              ColorSelector(),
            ],
          ),
          bottomNavigationBar: NavigationBar(
            destinations: const [
              NavigationDestination(
                icon: Icon(Icons.color_lens),
                label: 'Accent',
              ),
              NavigationDestination(
                icon: Icon(Icons.star),
                label: 'Flavor',
              ),
              NavigationDestination(
                icon: Icon(Icons.sentiment_neutral),
                label: 'Neutral',
              ),
              NavigationDestination(
                icon: Icon(Icons.colorize),
                label: 'Custom',
              )
            ],
            selectedIndex: _currentPage,
            onDestinationSelected: (index) =>
                setState(() => _currentPage = index),
          ),
        );
      }),
    );
  }
}

class ColorView extends StatelessWidget {
  const ColorView({Key? key, required this.colors}) : super(key: key);

  final Map<String, MaterialColor> colors;

  @override
  Widget build(BuildContext context) {
    return GridView(
      primary: false, // https://github.com/flutter/flutter/issues/93862
      padding: const EdgeInsets.all(tileSpacing),
      gridDelegate: SliverGridDelegateWithMaxCrossAxisExtent(
        maxCrossAxisExtent: tileWidth,
        mainAxisExtent: tileHeight * shades.length,
        mainAxisSpacing: tileSpacing,
        crossAxisSpacing: tileSpacing,
      ),
      children: colors.entries.map((c) {
        return ColorSwatch(name: c.key, color: c.value);
      }).toList(),
    );
  }
}

class ColorSelector extends StatefulWidget {
  const ColorSelector({Key? key}) : super(key: key);

  @override
  State<ColorSelector> createState() => _ColorSelectorState();
}

class _ColorSelectorState extends State<ColorSelector> {
  var _currentColor = YaruColors.generate(0xFF77216F);
  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        const Spacer(),
        Column(
          children: [
            const Spacer(),
            ColorPicker(
              color: _currentColor,
              enableShadesSelection: false,
              pickersEnabled: const {
                ColorPickerType.accent: false,
                ColorPickerType.primary: false,
                ColorPickerType.custom: false,
                ColorPickerType.wheel: true,
              },
              onColorChanged: (color) {
                setState(
                    () => _currentColor = YaruColors.generate(color.value));
              },
            ),
            const Spacer(),
          ],
        ),
        const Spacer(),
        SizedBox(
          width: tileWidth,
          child: Padding(
            padding: const EdgeInsets.symmetric(horizontal: 24),
            child: ColorSwatch(name: 'custom', color: _currentColor),
          ),
        ),
        const Spacer(),
      ],
    );
  }
}

class ColorSwatch extends StatelessWidget {
  const ColorSwatch({
    Key? key,
    required this.name,
    required this.color,
  }) : super(key: key);

  final String name;
  final MaterialColor color;

  @override
  Widget build(BuildContext context) {
    return Column(
      key: Key(name),
      mainAxisSize: MainAxisSize.min,
      children: shades.map((shade) {
        return ColorTile(
          name: name,
          shade: shade,
          color: color[shade]!,
          isPrimary: color.value == color[shade]!.value,
        );
      }).toList(),
    );
  }
}

class ColorTile extends StatelessWidget {
  const ColorTile({
    Key? key,
    required this.name,
    required this.shade,
    required this.color,
    required this.isPrimary,
  }) : super(key: key);

  final String name;
  final int shade;
  final Color color;
  final bool isPrimary;

  @override
  Widget build(BuildContext context) {
    return DefaultTextStyle(
      style: TextStyle(
        fontFamily: 'Ubuntu Mono',
        fontSize: Theme.of(context).textTheme.bodySmall!.fontSize,
        color: color.isDark ? YaruColors.porcelain : YaruColors.textGrey,
        fontWeight: isPrimary ? FontWeight.bold : FontWeight.normal,
      ),
      child: Container(
        padding: const EdgeInsets.all(8),
        color: color,
        height: tileHeight,
        child: Row(
          children: [
            Expanded(
              child: Text(
                isPrimary ? name : '$name[$shade]',
                overflow: TextOverflow.ellipsis,
              ),
            ),
            const SizedBox(width: 8),
            SelectableText(color.toHex()),
          ],
        ),
      ),
    );
  }
}

extension ColorX on Color {
  bool get isDark {
    return ThemeData.estimateBrightnessForColor(this) == Brightness.dark;
  }

  String toHex() {
    return '#${(value & 0xffffff).toRadixString(16).padLeft(6, '0')}';
  }
}
